### PR TITLE
Dev sync project page

### PR DESCRIPTION
**Pull request by kkjamie**
Known issues:
* Tasks: if one fails the whole chain fails. Ok in most situations but on a sync, if one package fails we kinda wanna keep going on the others.
* If a package failed to sync, it would be nice to inform the user, and how they can go fix it manually.
* Output: Git tasks often output to std_err, at the moment we pipe all that into the console. it's not ideal.
* The sync action, is a compound or epic action made up of other sub actions. It only dispatches on the sub action (CompiliePackageStatus), we should probably manipulate some state to show progress during the whole task.
* There is no state change in the store to indicate the results of the sync (right now it always assumes it completed, theres no concept of success fail etc.) the user feedback for that is all console message, and it's pretty noisy.

### CODE REVIEW

File: Editor/Tasks/AbstractProcessTask.cs [hunk 1]
```diff
@@ -9,17 +11,27 @@ internal abstract class AbstractProcessTask : ITask
 	{
 		private static readonly string projectDirectory;
 
+		public bool ReceivedError
```
Comment (borntocompile) [inline Editor/Tasks/AbstractProcessTask.cs:14]: `RecievedError` sounds like a method name, maybe `DidReceiveError` or `HasError` or `DidError`

File: Editor/Tasks/AbstractProcessTask.cs [hunk 2]
```diff
@@ -30,40 +42,42 @@ public void Execute(Action onComplete = null)
 
 		protected void RunProcessWithArgs(string fileName, string args)
 		{
-			Debug.Log("Executing:: " + args);
-
-			var process = new Process();
+			Debug.Log("Executing:: " + args + "\r\nFrom " + WorkingDirectory);
```
Comment (borntocompile) [inline Editor/Tasks/AbstractProcessTask.cs:45]: Minor: is the double colon intentional?

**Review by LukeAvery92 — APPROVED** [review]
There are minor errors with spaces and names but overall this PR is of good quality and completes the feature as expected.

File: Editor/UI/Actions.cs [hunk 3]
```diff
@@ -136,6 +131,101 @@ public static void SwitchVersion(AvailablePackage package, string version)
...
+						var version = packageStatus.RequiredVersion;
+						var packageName = packageStatus.PackageName;
+						var gitUrl = packageStatus.GitUrl;
+						var relativeInstallDirectory = Settings.RelativePackagesDirectoryPath + packageName;
+						var absoluteInstallDirectory = Settings.AbsolutePackagesDirectoryPath + packageName;
+
+						if (packageStatus.IsMissing)
+						{
+							taskChain.Add(
+								new AddSubmoduleTask(gitUrl, relativeInstallDirectory));
```
Comment (borntocompile) [inline Editor/UI/Actions.cs:154]: Minor: not keen on `new AddSub...` being on a new line

Comment (borntocompile) [reply Editor/UI/Actions.cs:154]: Same for the rest of 'em

File: Editor/UI/Actions.cs [hunk 4]
```diff
@@ -136,6 +131,101 @@ public static void SwitchVersion(AvailablePackage package, string version)
...
+						else if (packageStatus.IsOnWrongVersion)
+						{
+							taskChain.Add(
+								new CheckoutSubmoduleTask(
+									absoluteInstallDirectory,
+									version));
+						}
+					}
+
+					taskChain.Execute(() =>
```
Comment (borntocompile) [inline Editor/UI/Actions.cs:167]: Minor: you can do `taskChain.Execute(() => Debug.Log("Synced Project."));` 
I think it reads nicer

File: Editor/UI/Actions.cs [hunk 5]
```diff
@@ -136,6 +131,101 @@ public static void SwitchVersion(AvailablePackage package, string version)
...
+			tasks.Execute(() =>
+			{
+				Debug.Log("Project package status: ");
+
+				foreach (var package in packageListStatus.Packages)
+				{
+					Debug.Log(package);
+				}
+
+				Debug.Log("The project is " +
```
Comment (borntocompile) [inline Editor/UI/Actions.cs:220]: Don't think you need the line breaks here, there are lines much longer than this in other classes that don't break.

File: Editor/UI/Stores/PackageStore.cs [hunk 6]
```diff
@@ -125,5 +128,20 @@ private void WritePackagesJson()
 			var jsonText = JsonUtility.ToJson(InstalledPackages.Value, true);
 			File.WriteAllText(Settings.AbsolutePackagesJsonFilePath, jsonText);
 		}
+
+		private void HandleCompletPackageListStatusStarted(Action obj)
```
Comment (borntocompile) [inline Editor/UI/Stores/PackageStore.cs:132]: Typo: `HandleCompletPackageListStatusStarted` complete

File: Editor/UI/Stores/PackageStore.cs [hunk 7]
```diff
@@ -125,5 +128,20 @@ private void WritePackagesJson()
...
+		{
+			PackageListStatus.SetValue(null);
+			IsWorking.SetValue(true);
+			Operation.SetValue("Compiling status of each package");
+		}
+
+		private void HandleCompletPackageListStatusComplete(Action obj)
```
Comment (borntocompile) [inline Editor/UI/Stores/PackageStore.cs:139]: Wouldn't you prefer to call `obj` `action` instead?

File: Editor/Tasks/AbstractProcessTask.cs (hunk 1 above)
Comment (kkjamie) [reply Editor/Tasks/AbstractProcessTask.cs:14]: Agreed although to me it's not a method name, ReceiveError() would be. I already changed it on the next branch

File: Editor/Tasks/AbstractProcessTask.cs (hunk 2 above)
Comment (kkjamie) [reply Editor/Tasks/AbstractProcessTask.cs:45]: Yeah but I can change it if you think it needs it

File: Editor/UI/Stores/PackageStore.cs (hunk 7 above)
Comment (kkjamie) [reply Editor/UI/Stores/PackageStore.cs:139]: Yeah, I didn't choose the name, rider auto generated. I'll changed

Comment (kkjamie) [reply Editor/UI/Stores/PackageStore.cs:139]: I have changed it to `action`. Although you should be aware this is not a `System.Action` delegate. It's a flux action, ie: a small object describing changes in state

**Comment by kkjamie** [comment]
I'm closing this, so much has changed in the next branch that it would be unconstructive to merge it. It will make the next PR much bigger however

**Pull request closed without merging**